### PR TITLE
dnn: better net exporter that works with netron

### PR DESCRIPTION
**Pull request by fengyuentau**
Current net exporter `dump` and `dumpToFile` exports the network structure (and its params) to a .dot file which works with `graphviz`. This is hard to use and not friendly to new user. What's worse, the produced picture is not looking pretty.

This PR introduces new exporter `dumpToPbtxt` and uses this new exporter by default with environment variable `OPENCV_DNN_NETWORK_DUMP`. It mimics the string output of a onnx model but modified with dnn-specific changes, see below for an example.

image

## Usage

Call `cv::dnn::Net::dumpToPbtxt`:

```cpp
TEST(DumpNet, dumpToPbtxt) {
    std::string path = "/path/to/model.onnx";
    auto net = readNet(path);

    Mat input(std::vector<int>{1, 3, 640, 480}, CV_32F);
    net.setInput(input);

    net.dumpToPbtxt("yunet.pbtxt");
}
```

Set `export OPENCV_DNN_NETWORK_DUMP=1`

```cpp
TEST(DumpNet, env) {
    std::string path = "/path/to/model.onnx";
    auto net = readNet(path);

    Mat input(std::vector<int>{1, 3, 640, 480}, CV_32F);
    net.setInput(input);

    net.forward();
}
```

---

Note:
- `pbtxt` is registered as one of the ONNX model suffix in netron. So you can see `module: ai.onnx` and such in the model.
- We can get the string output of an ONNX model with the following script

```python
import onnx
net = onnx.load("/path/to/model.onnx")
net_str = str(net)
file = open("/path/to/model.pbtxt", "w")
file.write(net_str)
file.close()
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by fengyuentau** [comment]
Here is one example I get from this patch [models.zip](https://github.com/opencv/opencv/files/15305984/models.zip). They all can be opened with Netron.

**Note that I did not test other models.**

**Comment by asmorkalov** [comment]
I propose to add dump call to `apps/model-diagnostics` to get graph every time, when the network was loaded or do it by option.

File: modules/dnn/src/net_impl.cpp [hunk 1]
```diff
@@ -1830,15 +1830,266 @@ string Net::Impl::dump(bool forceAllocation) const
...
+    // Add nodes, inputs and outputs
+    for (std::map<int, LayerData>::const_iterator iter = map.begin(); iter != map.end(); iter++) {
+        auto &ld = iter->second;
+        if (ld.id == 0) {
+            for (int i = 0; i < ld.outputBlobs.size(); i++) {
+                const auto &name = netInputLayer->outNames.empty() ? cv::format("%s_%d", ld.name.c_str(), i) : netInputLayer->outNames[i];
+                out << indent_spaces << "input {\n"
+                    << indent_spaces << format("  name: \"%s\"\n", name.c_str());
+                // Add shape
+                if (ld.outputBlobs.empty()) {
```
Comment (fengyuentau) [inline modules/dnn/src/net_impl.cpp:2008]: Found a typo here. Will fix it tomorrow.

File: modules/dnn/src/net_impl.cpp [hunk 2]
```diff
@@ -1830,15 +1830,266 @@ string Net::Impl::dump(bool forceAllocation) const
     return out.str();
 }
 
+static void dumpShapeToString(std::ostringstream &out, const Mat &m, const int num_indent_spaces = 4) {
```
Comment (dkurt) [inline modules/dnn/src/net_impl.cpp:1833]: Not only shape. Maybe `dumpTensorToString`?

File: modules/dnn/src/net_impl.cpp [hunk 3]
```diff
@@ -1830,15 +1830,266 @@ string Net::Impl::dump(bool forceAllocation) const
...
+        case CV_8U:  type = 2; break;
+        case CV_8S:  type = 3; break;
+        case CV_16U: type = 4; break;
+        case CV_16S: type = 5; break;
+        case CV_32S: type = 6; break;
+        // case CV_64S: type = 7; break;
+        // case CV_STR: type = 8; break;
+        // case CV_BOOL: type = 9; break;
+        case CV_16F: type = 10; break;
+        // don't use default
```
[the remark this answers is not in the document]
Comment (fengyuentau) [reply modules/dnn/src/net_impl.cpp:1849]: Macro `OPENCV_VERSION_MAJOR` is not available in `net_impl.cpp`.

File: modules/dnn/src/net_impl.cpp (hunk 2 above)
Comment (fengyuentau) [reply modules/dnn/src/net_impl.cpp]: You are right. Renamed to `dumpTensorToString`.

**Comment by fengyuentau** [comment]
> I propose to add dump call to `apps/model-diagnostics` to get graph every time, when the network was loaded or do it by option.

Unfortunately it cannot be added to `apps/model-diagnostics`. I took a look at the app, and found it only reads the net but does not call `net.setInput()`, which needs to be called first in order to get the graph.

However, we can add option in the app asking for the input dimension and then we can call the new exporter to get the graph. Is it ok if I add this option?

**Comment by asmorkalov** [comment]
@fengyuentau Let's extend the tool or add the call to some generic sample. It'll be great to have the option available without codding. It helps to triage incoming issues.

**Comment by fengyuentau** [comment]
> @fengyuentau Let's extend the tool or add the call to some generic sample. It'll be great to have the option available without codding. It helps to triage incoming issues.

I mean we can extend the tool only after we figure out how to set the input for the net. Current `CommandLineParser` does not satisfy our needs to have inputs unless we code everytime with the input of the model.

> add the call to some generic sample

Do you mean existing ones in `samples/dnn/` ?

**Comment by asmorkalov** [comment]
I propose to add several cmdline options line `--input_shape1`, `--input_shape2`, ... `--input_shape5` and parse formula like 128x128x3 with sscanf. It should be enough  to define inputs without CommadLineParser modification.

**Comment by fengyuentau** [comment]
> I propose to add several cmdline options line `--input_shape1`, `--input_shape2`, ... `--input_shape5` and parse formula like 128x128x3 with sscanf. It should be enough to define inputs without CommadLineParser modification.

Ok. Although this is not a perfact solution, it can satisfy most of our needs.

**Comment by fengyuentau** [comment]
> I propose to add several cmdline options line `--input_shape1`, `--input_shape2`, ... `--input_shape5` and parse formula like 128x128x3 with sscanf. It should be enough to define inputs without CommadLineParser modification.

@asmorkalov It is implemented. It works as follows

```shell
./build/bin/opencv_model_diagnostics -model=/workspace/opencv_zoo/models/face_detection_yunet/face_detection_yunet_2023mar.onnx -input0_name=input -input0_shape=1,3,640,640
```

File: modules/dnn/src/net_impl.cpp [hunk 4]
```diff
@@ -1830,15 +1830,270 @@ string Net::Impl::dump(bool forceAllocation) const
...
+        case CV_16U: type = 4; break;
+        case CV_16S: type = 5; break;
+        case CV_32S: type = 6; break;
+/*
+#if OPENCV_VERSION_MAJOR > 4
+        case CV_64S: type = 7; break;
+        case CV_STR: type = 8; break;
+        case CV_BOOL: type = 9; break;
+#endif
+*/
```
Comment (asmorkalov) [inline modules/dnn/src/net_impl.cpp:1851]: Please uncomment it.

Comment (fengyuentau) [reply modules/dnn/src/net_impl.cpp:1851]: `OPENCV_VERSION_MAJOR` is undefined in `net_impl.cpp`. It is not going to work unless it is defined in 5.x.

Comment (asmorkalov) [reply modules/dnn/src/net_impl.cpp:1851]: you just need to include "opencv2/core/version.hpp"

Comment (fengyuentau) [reply modules/dnn/src/net_impl.cpp:1851]: The macro name is actually `CV_VERSION_MAJOR`, it is available already.

**Pull request merged**